### PR TITLE
dialects: Add SingleBlockImplicitTerminator trait to dialects

### PR DESCRIPTION
**Pull request by compor**
This PR adds `SingleBlockImplicitTerminator` trait to dialects without calling `ensure_terminator` which is to be used for custom printing/parsing.

This added to:
- `scf`
- `gpu`

For `scf` the operation definitions had to be reordered due to #1218.

### CODE REVIEW

**Comment by math-fehr** [comment]
Seems like the stack failed :/

**Comment by compor** [comment]
> Seems like the stack failed :/

what do you mean?
those are the expected failures coming from the parser, but I'll address as we discussed on Zulip.

**Comment by math-fehr** [comment]
I mean that the github PR stack seems to have failed. Currently, your PR is targeting main, but you still have changes that add SingleBlockImplicitTerminator class.

File: tests/test_rewriter.py [hunk 1]
```diff
@@ -3,10 +3,9 @@
 import pytest
 from conftest import assert_print_op
 
+from xdsl.dialects import test
 from xdsl.dialects.arith import Addi, Arith, Constant
 from xdsl.dialects.builtin import Builtin, ModuleOp, i32, i64
-from xdsl.dialects.func import Func
```
Comment (compor) [inline tests/test_rewriter.py:8]: hmm, maybe a separate PR for these changes too?

**Review by math-fehr — APPROVED** [review]
Nice!

File: xdsl/dialects/scf.py [hunk 2]
```diff
@@ -205,20 +254,11 @@ def verify_(self) -> None:
...
-            )
-
         # Ensure that the number of operands in scf.yield is exactly zero
-        number_yield_ops = len(self.body.block.last_op.arguments)
-        if number_yield_ops != 0:
+        if (last_op := self.body.block.last_op) is not None and last_op.operands:
             raise VerifyException(
-                f"scf.yield contains {number_yield_ops} operands but this must be zero inside an scf.parallel"
+                f"scf.yield contains {len(last_op.operands)} operands but this must be "
+                "0 inside an scf.parallel"
```
[the remark this answers is not in the document]
Comment (compor) [reply xdsl/dialects/scf.py:261]: I don't think I changed it, but I'm in for improving it like this. Thanks!

**Review by superlopuh — APPROVED** [review]
LGTM

**Review by webmiche — APPROVED** [review]
LGTM!

**Review by PapyChacal — APPROVED** [review]

**Pull request merged**